clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionSimulationCoordinator.java

```diff
@@ -0,0 +1,316 @@
+import org.apache.kafka.common.requests.AddOffsetsToTxnRequest;
+import org.apache.kafka.common.requests.AddOffsetsToTxnResponse;
+import org.apache.kafka.common.requests.AddPartitionsToTxnRequest;
+import org.apache.kafka.common.requests.AddPartitionsToTxnResponse;
+import org.apache.kafka.common.requests.EndTxnRequest;
+import org.apache.kafka.common.requests.EndTxnResponse;
+import org.apache.kafka.common.requests.FindCoordinatorRequest;
+import org.apache.kafka.common.requests.FindCoordinatorResponse;
+import org.apache.kafka.common.requests.InitProducerIdRequest;
+import org.apache.kafka.common.requests.InitProducerIdResponse;
+import org.apache.kafka.common.requests.ProduceRequest;
+import org.apache.kafka.common.requests.ProduceResponse;
+import org.apache.kafka.common.requests.ProduceResponse.PartitionResponse;
+import org.apache.kafka.common.requests.TransactionResult;
+import org.apache.kafka.common.requests.TxnOffsetCommitRequest;
+import org.apache.kafka.common.requests.TxnOffsetCommitResponse;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Random;
+
+import static org.apache.kafka.common.record.RecordBatch.NO_PRODUCER_EPOCH;
+import static org.apache.kafka.common.record.RecordBatch.NO_PRODUCER_ID;
+
+/**
+ * A reduced functionality of a combination of transaction coordinator and group coordinator.
+ * It provides basic event handling from {@link Sender} with transaction turned on.
+ *
+ * Random fault injection is supported as well, which will return a retriable error to
+ * the client.
+ */
+class TransactionSimulationCoordinator {
+
+    private final Map<TopicPartition, List<Record>> pendingPartitionData;
+    private final Map<TopicPartition, Long> pendingOffsets;
+    private boolean offsetsAddedToTxn = false;
+
+    private long currentProducerId = 0L;
+    private short currentEpoch = 0;
+    private Random faultInjectRandom = new Random();
+
+    public Map<TopicPartition, List<Record>> persistentPartitionData() {
+        return persistentPartitionData;
+    }
+
+    public Map<TopicPartition, Long> committedOffsets() {
+        return committedOffsets;
+    }
+
+    private final Map<TopicPartition, List<Record>> persistentPartitionData;
+    private final Map<TopicPartition, Long> committedOffsets;
+
+    private final MockClient networkClient;
+    private final int throttleTimeMs = 10;
+
+    TransactionSimulationCoordinator(MockClient networkClient) {
+        this.networkClient = networkClient;
+        this.pendingPartitionData = new HashMap<>();
+        this.pendingOffsets = new HashMap<>();
+        this.persistentPartitionData = new HashMap<>();
+        this.committedOffsets = new HashMap<>();
+    }
+
+    void runOnce(boolean dropMessage) {
+        Queue<ClientRequest> incomingRequests = networkClient.requests();
+
+        final boolean faultInject = faultInjectRandom.nextBoolean();
```

Comment:
Wondering if it would be more useful if we can control the faults more explicitly. For example, we could add a hook to make the coordinator temporarily unavailable and to restore it later.